### ISSUE TITLE
[ENH]: subfigure_mosaic

### ISSUE THREAD

**Issue by anntzer** [matplotlib/matplotlib#25949]
### Problem

Recently I had a case of layouting subfigures where I could have benefitted from a subfigure_mosaic function, similar to subplot_mosaic.

### Proposed solution

Provide subfigure_mosaic.
Bonus points if the implementation shares as much as possible with subplot_mosaic (I suspect we can just factor out everything and pass the "child-adding method" (fig.add_subplot or fig.add_subfigure) as a parameter to the internal implementation?).

**Comment by tacaswell** [comment]
I think this is a good first issue (limited API discussion as "match subplot_mosaic" is going to be the answer) and only affects a very limited part of the codebase.

Labeling this as medium difficluty because the subplot_mosaic code is a bit complex (it supports infinite nesting of layouts via recursion) and some of the internal variables are not super clearly named (I named them).  

The source is:

https://github.com/matplotlib/matplotlib/blob/82cfed34e2a54fcf450716deb6932af0c151395e/lib/matplotlib/figure.py#L1821-L2128

On a cursory skim, I think that https://github.com/matplotlib/matplotlib/blob/82cfed34e2a54fcf450716deb6932af0c151395e/lib/matplotlib/figure.py#L2081-L2087 is the only Axes-specific thing and parameterizin `_do_layout` to take `make_the_sub_thing` as an input would work.

One maybe interesting question is that for nested layouts to we want to have sub-figures of sub-figures or flatten all of the sub-figures to be children of the Figure they are called on?  I think the return has to be flat so I would lean towards the implementation also being flat (with the nesting in the spec being a details of the layout not an expression of actual nesting) and if the user _does_ really want sub-figure of sub-figure they can call `subfigure_moasic` in a loop to build up exactly the tree they want.

Tasks:

 - understand how `subplot_mosaic` currently works
 - refactor the layout logic into a private method
 - pass the "make the thing" method through `_do_layout`
 - add new FigureBase method `subfigure_mosaic`
 - tests (lots of tests....should be able to adapt the subplot_mosaic tests)
 - whats new entry (suspect that this is something  people will be excited about so we want to highlight it)

software estimate are notoriously hard, but I would guess 4-20 hours with a bulk of the time being in writing tests and docs.

**Comment by anntzer** [comment]
> One maybe interesting question is that for nested layouts to we want to have sub-figures of sub-figures or flatten all of the sub-figures to be children of the Figure they are called on? I think the return has to be flat

Flat would seem normal to me.

**Comment by brandall25** [comment]
Could I have a go at this?

**Comment by melissawm** [comment]
Hi @brandall25 - go ahead! Make sure you check our [contributor guide](https://matplotlib.org/stable/devel/index) and let us know if you have questions. Thanks!

**Comment by tacaswell** [comment]
> When you say "refactor the layout logic into a private method", do you want that to be a private method of FigureBase that both subplot_mosaic and subfigure_mosaic can use?

yes, or a class method,  top level (private) function if it does not need any state.  The main goal is to not have two copies of this somewhat long subtle code in the code base.

> Would this be add_subplot for subplot_mosaic and maybe add_subfigure for subfigure_mosaic?

Yes, exactly that.  The logic is "I need to add a sub-version of <what ever I am laying out>" which we can specify via dependency injection.

**Comment by turnipseason** [comment]
@brandall25 Hi! Not sure I'm qualified to answer you here, but I've been working on this since last week and did just what you described for both of your questions. It seems to work.
The only problem now is writing proper docs and tests, because just passing a string with no kwargs for `subfigure_mosaic` doesn't create anything pretty like the plot function does. In fact, it looks like nothing at all. That's probably intended (since it does work if you add kwargs), but still.

**Comment by brandall25** [comment]
@turnipseason Hello, thanks for responding! Sounds like you have it covered, I think I'll move on to another issue.

**Comment by turnipseason** [comment]
@brandall25 Hi! Sorry, I really should've written a comment earlier. Good luck with the other issues!

**Comment by tacaswell** [comment]
@turnipseason Yes please to the PR.

Adding to or mimicing https://github.com/matplotlib/matplotlib/blob/main/galleries/users_explain/axes/mosaic.py for sub-figures would probably make sense.  You can use the same `fig.text(.5, .5, LABEL, ha='center', va='center')` (in fact if you pass a `Figure` object to `identify_axes` I think it will "just work"... 🦆 typing is fun when it works :)

Alternatively looking https://matplotlib.org/devdocs/gallery/subplots_axes_and_figures/subfigures.html#figure-subfigures and using the "change the patch color" as a way to label the space (or both!) may be a good starting point for documentation inspiration. 

Similarly look at how `subplot_mosaic` and SubFigure are tested, I suspect they can be adapted.

**Comment by tacaswell** [comment]
also don't worry about the first draft being perfect, we expect iteration on PRs.

**Comment by klorine28** [comment]
Is the issue still open, or has the issue changed it would seem to me from the comment thread it has been resolved.

**Comment by tacaswell** [comment]
@klorine28 There is an active PR addressing this issue (see cross reference right above your comment), I suggest you look for a different issue to work on.

**Comment by saikarna913** [comment]
@tacaswell is this issue still exits

**Comment by mgamzec** [comment]
Is the issue still open?

**Comment by melissawm** [comment]
Hi @mgamzec - as mentioned above there is still a PR open for this although it hasn't seen movement in a while. In the meantime, please try and pick another issue to work on. Thanks!

**Comment by JoshTVR** [comment]
Hi, I see this was already implemented by @turnipseason. 
Should this issue be closed?

Thanks!

**Comment by timhoffm** [comment]
No. The PR (#26061) is not merged.

**Comment by p-sreesankar** [comment]
Hello, Can I have a go at this issue.

**Comment by timhoffm** [comment]
#26061 is still work in progress. Please do not open a concurrent pull request. If you want to contribute here, check the status of #26021 and try help clarifying the state / comment on what still needs to be done.

Given there is already a PR, and the length of its discussion I'm removing "good first issue".

On a side note, I don't think `subfigure_mosaic` is really necessary as the need for complex subfigure layouts should be very rare. I won't get involved here and leave it to the original proponents to decide the future of this topic.

**Comment by story645** [comment]
> , I don't think subfigure_mosaic is really necessary as the need for complex subfigure layouts should be very rare

So I've been really wanting it lately cause I use subfigures a lot for reusing panels across paper/poster/slide deck.
